tsdb/index/tsi1/file_set.go

```diff
@@ -30,6 +30,14 @@ func NewFileSet(levels []CompactionLevel, sfile *tsdb.SeriesFile, files []File)
 	}, nil
 }
 
+func (fs *FileSet) Clone() *FileSet {
+	fs2 := &FileSet{
+		files: make([]File, len(fs.files)),
+	}
+	copy(fs2.files, fs.files)
```

Comment:
I'm not sure this is safe.  `File` is an interface and the value of the interface can contain pointers.

It looks like `LogFile` and `IndexFile` implement `File` and both of those types contain mutexes, pointers, etc.

so those pointers and the contents of the mutexes will be copied.

are we sure this is thread safe?  should be be copying mutex state?  is it okay to copy the pointers?

here's a small program that demonstrates the problems:

```package main

import (
  "log"
  "math/rand"
)

type File interface {
  Open() (int, error)
}

type F struct {
  m map[string]int
  n *int
}

func (f F) Open() (int, error) {
  log.Printf("%p:%p", &f, f.n)
  return 0, nil
}

func main() {
  var files []File
  for i := 0; i < 3; i++ {
    v := rand.Intn(100)
    files = append(files, F{n: &v, m: make(map[string]int)})
  }

  for _, v := range files {
    v.Open()
  }
  newFiles := make([]File, len(files))

  copy(newFiles, files)

  for _, v := range newFiles {
    v.Open()
  }
}
```

if you agree that this is a problem, how would you fix it?

what if each file also had a clone method that returns a safely copied version of itself -- without copying mutexes and maybe deeply copying pointer values?

i'm not sure but this seems like a potential source of bugs.